operator/controllers/remote.go

```diff
@@ -0,0 +1,187 @@
+
+	if client.IgnoreNotFound(err) != nil {
+		// unknown error
+		return nil, nil, err
+	} else if err != nil {
+		// remote resource doesn't exist
+		remoteManifest, err = remoteInterface.createRemote(ctx)
+		if err != nil {
+			return nil, nil, err
+		}
+	}
+
+	// check finalizer
+	if !controllerutil.ContainsFinalizer(remoteManifest, labels.ManifestFinalizer) {
+		controllerutil.AddFinalizer(remoteManifest, labels.ManifestFinalizer)
+		if err = remoteInterface.Update(ctx, remoteManifest); err != nil {
+			return nil, nil, err
+		}
+	}
+
+	return remoteInterface, remoteManifest, nil
+}
+
+func (r *RemoteInterface) createRemote(ctx context.Context) (*v1alpha1.Manifest, error) {
+	remoteManifest := &v1alpha1.Manifest{}
+	remoteManifest.Name = r.NativeObject.Name
+	remoteManifest.Namespace = r.NativeObject.Namespace
+	remoteManifest.Spec = *r.NativeObject.Spec.DeepCopy()
+	controllerutil.AddFinalizer(remoteManifest, labels.ManifestFinalizer)
+	err := r.CustomClient.Create(ctx, remoteManifest)
+	return remoteManifest, err
+}
+
+func (r *RemoteInterface) SpecSyncToNative(ctx context.Context, remoteManifest *v1alpha1.Manifest) (bool, error) {
+	if remoteManifest.Status.ObservedGeneration == 0 {
+		// observed generation missing - remote resource freshly installed!
+		return true, nil
+	}
+
+	if remoteManifest.Status.ObservedGeneration != remoteManifest.Generation {
+		// outdated
+		r.NativeObject.Spec = *remoteManifest.Spec.DeepCopy()
+		remoteManifest.Status.ObservedGeneration = remoteManifest.Generation
+		return false, r.UpdateStatus(ctx, remoteManifest)
+	}
+
+	return true, nil
+}
+
+func (r *RemoteInterface) StateSyncToRemote(ctx context.Context, remoteManifest *v1alpha1.Manifest) error {
+	if r.NativeObject.Status.State != remoteManifest.Status.State {
+		remoteManifest.Status.State = r.NativeObject.Status.State
+		remoteManifest.Status.ObservedGeneration = remoteManifest.Generation
+		return r.UpdateStatus(ctx, remoteManifest)
+	}
+	return nil
+}
+
+func (r *RemoteInterface) SpecSyncToRemote(ctx context.Context, remoteManifest *v1alpha1.Manifest) (*v1alpha1.Manifest, error) {
+	// ignore cases where native object has never been reconciled!
+	if r.NativeObject.Status.ObservedGeneration == 0 {
+		return remoteManifest, nil
+	}
+
+	if r.NativeObject.Status.ObservedGeneration != r.NativeObject.Generation {
+		// remove finalizer
+		if err := r.RemoveFinalizerOnRemote(ctx, remoteManifest); err != nil {
+			return remoteManifest, err
+		}
+
+		// trigger delete on remote
+		if err := r.HandleDeletingState(ctx, remoteManifest); err != nil {
+			return remoteManifest, err
+		}
+
+		// create fresh object
+		return r.createRemote(ctx)
+	}
+
+	return remoteManifest, nil
+}
+
+func (r *RemoteInterface) CreateCRD(ctx context.Context) error {
+	crd := v1.CustomResourceDefinition{}
+	err := r.CustomClient.Get(ctx, client.ObjectKey{
+		// TODO: Change "manifests" with updated api value
+		Name: fmt.Sprintf("%s.%s", "manifests", v1alpha1.GroupVersion.Group),
```

Comment:
dont you have a better alternative now available in types with the plural available?

Reply:
Yes I do, but for that the api folder should be a part of the repo. This will be a part of a follow-up PR, thus the TODO :)